### DISCUSSION TITLE
Title autogeneration problem

### DISCUSSION THREAD

**Discussion by unameisfine** [AllYourBot/hostedgpt#497] [General]
I've run into a problem: my automatic chat titles aren't being generated. Can anyone suggest how to debug this issue? I've checked the logs but found no relevant entries. I've reviewed the code and the trigger conditions, which logically should be firing. The titles in the database are empty, and the dialogue has an API. Could this be related to my use of the gpt4o model? I've seen messages that gpt 3.5 is typically used for this task.

I've already checked the trigger conditions in the code, database entries, and confirmed API availability for the dialogue. Any ideas on what else I should look into or how to troubleshoot this further? Thanks in advance for any help or insights!

**Comment by keithschacht** [comment]
Hey, that's really odd. It should be working. It does require an OpenAI API key. I have a task to make autogeneration work with any of the API backends but it doesn't yet—currently it always tries to use OpenAI. But it sounds like you have that. I just tried my own instance of the app to make sure there isn't some recent breakage. Do you know how to get to a Rails console in your setup? If so, try this out for me (customize the email to your login email):

```
> Current.user = Person.with_personable_type.find_by(email: 'YOUR_LOGIN@email.com').user; AutotitleConversationJob.new.send(:generate_title_for, "this is a long message in which we're talking about video games. zelda is great but angry birds is nice too")
=> "Video games discussion"
```
If this works, the core method is communicating with the AI. Then try (update email again):
```
Conversation.last.update!(title: nil); puts "### Old: #{Conversation.last.title}"; Current.user = Person.with_personable_type.find_by(email: 'YOUR_LOGIN@email.com').user; AutotitleConversationJob.new.perform(Conversation.last.id); puts "### New: #{Conversation.last.title}"
```
Scroll back through the history and find ### Old and ### New. If this works then it means the job successfully works, but maybe it's a queue issue? Try this to run the queue (update email again):
```
Conversation.last.update!(title: nil); puts "### Old: #{Conversation.last.title} changed at #{Conversation.last.updated_at}"; sleep 2; Current.user = Person.with_personable_type.find_by(email: 'YOUR_LOGIN@email.com').user; AutotitleConversationJob.perform_later(Conversation.last.id); sleep 5; puts "### New: #{Conversation.last.title} changed at #{Conversation.last.updated_at}"
```
This time make sure that the title changed from empty (old) to new again and also make sure the timestamp changed by a few seconds.

Let me know the results of this and that will help to pinpoint the problem.

**Reply by unameisfine** [reply]
After 

` AutotitleConversationJob.new.perform(conversation.id)`

 I got error:
```
Faraday::UnauthorizedError: the server responded with status 401
from /usr/local/bundle/gems/faraday-2.8.1/lib/faraday/response/raise_error.rb:18:in `on_complete'
```

Full log:
```
 Conversation Load (0.2ms)  SELECT "conversations".* FROM "conversations" WHERE "conversations"."id" = $1 LIMIT $2  [["id", 372], ["LIMIT", 1]]
  Assistant Load (0.4ms)  SELECT "assistants".* FROM "assistants" WHERE "assistants"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  LanguageModel Load (0.3ms)  SELECT "language_models".* FROM "language_models" WHERE "language_models"."id" = $1 LIMIT $2  [["id", 4], ["LIMIT", 1]]
  APIService Load (0.3ms)  SELECT "api_services".* FROM "api_services" WHERE "api_services"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  Message Load (0.5ms)  SELECT "messages"."index", "messages"."version" FROM "messages" WHERE "messages"."conversation_id" = $1 ORDER BY "messages"."version" DESC, "messages"."index" DESC LIMIT $2  [["conversation_id", 372], ["LIMIT", 1]]
  Message Exists? (0.7ms)  SELECT 1 AS one FROM "messages" LEFT JOIN messages mB ON mB.branched_from_version = messages.version AND mB.conversation_id = messages.conversation_id AND mB.version <= 1 INNER JOIN messages mV ON mV.index = messages.index AND mV.conversation_id = messages.conversation_id WHERE "messages"."conversation_id" = $1 AND (messages.index <= 1 AND messages.version <= 1) LIMIT $2  [["conversation_id", 372], ["LIMIT", 1]]
  User Load (0.2ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 LIMIT $2  [["id", 1], ["LIMIT", 1]]
  Message Load (0.9ms)  SELECT DISTINCT ON (index) messages.*, CASE WHEN mB.version IS NOT NULL THEN mB.version ELSE messages.version END as max_branched_to, CASE WHEN messages.branched THEN STRING_AGG(mV.version::text, ',') OVER (PARTITION BY messages.index,messages.version ORDER BY mV.version ROWS BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) ELSE null END AS versions FROM "messages" LEFT JOIN messages mB ON mB.branched_from_version = messages.version AND mB.conversation_id = messages.conversation_id AND mB.version <= 1 INNER JOIN messages mV ON mV.index = messages.index AND mV.conversation_id = messages.conversation_id WHERE "messages"."conversation_id" = $1 AND (messages.index <= 1 AND messages.version <= 1) ORDER BY messages.index, "max_branched_to" DESC, messages.version DESC LIMIT $2  [["conversation_id", 372], ["LIMIT", 4]]
  APIService Load (0.4ms)  SELECT "api_services".* FROM "api_services" WHERE "api_services"."user_id" = $1 AND (api_services.deleted_at is null or api_services.deleted_at > '2024-08-26 16:20:53.989986') AND "api_services"."driver" = $2 AND (api_services.token is not null and length(api_services.token) > 0) ORDER BY "api_services"."id" ASC LIMIT $3  [["user_id", 1], ["driver", "openai"], ["LIMIT", 1]]
Faraday::UnauthorizedError: the server responded with status 401
from /usr/local/bundle/gems/faraday-2.8.1/lib/faraday/response/raise_error.rb:18:in `on_complete'
```

SQL Query 
```
 SELECT "api_services".* FROM "api_services" WHERE "api_services"."user_id" = 1
 AND (api_services.deleted_at is null or api_services.deleted_at > '2024-08-26 16:20:53.989986') 
 AND "api_services"."driver" = 'openai' 
 AND (api_services.token is not null and length(api_services.token) > 0) ORDER BY "api_services"."id" ASC LIMIT 1
```
return empty result because I have services without token. Condition "api_services.token is not null" filtered it. I'm using DEFAULT_OPENAI_KEY option in docker-compose

I don't wont to share my API keys with my friends whom I take access to my local instance.

**Reply by keithschacht** [reply]
Ohhhh, I know what the bug is. That helps!! Let me work on fixing this.

**Reply by keithschacht** [reply]
I tee'd up a PR for this but I ended up refactoring a lot of code. This was some needed refactoring. I think everything works (my tests are passing) but I need to do a bit more trying it out before I merge in. Hopefully I'll get to that tomorrow. In the meantime you are welcome to pull it down and try locally: https://github.com/AllYourBot/hostedgpt/pull/500

**Reply by keithschacht** [reply]
@unameisfine This is now fixed in main. I did the deeper fix so now whatever model you're using, even if it's just a local model, that's the model that is used to auto-generate the title.

Thanks for making me aware oft his issue!

**Reply by unameisfine** [reply]
Hello! I checked new version and got another error after "AutotitleConversationJob.new.perform(conversation.id)":

```
Connecting to OpenAI API server at https://api.openai.com/ with access token of length 51
JSON::ParserError: unexpected token at '```json
{ "topic": "Greetings and affairs" }
```'
from /usr/local/bundle/gems/json-2.7.1/lib/json/common.rb:219:in `parse'
```

Same error with Claude conversation:

```
Connecting to Anthropic API server at https://api.anthropic.com/ with access token of length 108
JSON::ParserError: unexpected token at '```json
{ "topic": "AI greeting" }
```'
from /usr/local/bundle/gems/json-2.7.1/lib/json/common.rb:219:in `parse'
```

**Reply by keithschacht** [reply]
Ugh. It can be challenging to get these AI models to reliably return JSON. I just pushed a hotfix that I think will it make it more resilient. Could you pull main and try it again?

**Reply by unameisfine** [reply]
Now it works! Thank you! It was an exciting quest 🖖
